codec: Raise max_request_headers_kb limit to 96 KiB

auni53 opened the pull request:
Signed-off-by: Auni Ahsan <auni@google.com>

Bump up max configurable `max_request_headers_kb` to 96 KiB.
Add a check to http1/codec_impl.cc for headers size.
Raise the default library limits in http_parser (at compilation) and nghttp2 (at options) to 0x7fffffff, so we'll rely on our own codec check.

*Risk Level*: Medium.
*Testing*: Moved all the large request headers tests to ProtocolIntegrationTest.

Part of #5626.

auni53 commented:
What's missing this PR is an actual runtime call of `http_parser_set_max_header_size`. That will go in the Http conn manager config, but it's impossible to test AFAIK because the proto validator limit is lower than the http_parser default limit. When the limit is bumped past 80kb, i'll add an http1 integration test and a settings call to conn manager/config.cc.

alyssawilk commented:
Hey Auni, as discussed offline, ping me when CI is happy?

/wait

auni53 commented:
Hi! I need some input on how to resolve this. Essentially, my test is failing because the nghttp2 library max request headers check, which I now manually configure, trips differently from the existing codec check. The codec check at codec_impl.cc:563 will check if the headers exceed the limit and reset the stream if so, while the nghttp2 library check will just flat-out close the stream.

How should I resolve this? Two ideas:
1. In the onStreamClose callback check if the nghttp2 limit was reached and interrupt with a reset stream if so. (Not sure if this is possible or logical.)
2. Set the nghttp2 limit to something arbitrarily higher and solely rely on our own check, whose behaviour we're in more control of. I was thinking either the max configurable limit from envoy; though since that would have to be edited twice each time the config limit is changed, maybe the configured limit + 1 would work.

Thoughts?

auni53 commented:
Ahh, some more details. The library check triggers before any header bytes are dispatched to the library, so server_callbacks_.newStream() is never called (and so it's never reset, which is why my test fails). So yeah again, I could configure the library check up high as I suggested to never get hit, or we could decide that behaviour is useful.

auni53 commented:
OK so I went with an arbitrarily hard-coded nghttp2 max-header-len setting of 100kb. This can be easily changed or be configurable later, but this changes no envoy behaviour. (Both before and after this PR, the nghttp2 max check will not ever be triggered.)

auni53 commented:
/retest

alyssawilk commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -853,6 +853,8 @@ ConnectionImpl::Http2Options::Http2Options(const Http2Settings& http2_settings)
   // of kept alive HTTP/2 connections.
   nghttp2_option_set_no_closed_streams(options_, 1);
   nghttp2_option_set_no_auto_window_update(options_, 1);
+  nghttp2_option_set_max_send_header_block_length(options_,
```
Isn't this the setting for the size of an individual header frame?  I think we want to change the total allowed headers, not the per-frame limits.

mattklein123 replied:
IIRC this is not very well named, but I think this does actually set the header send limit, and the library will chunk it into frames using continuation frames if necessary. @auni53 to check me here.

alyssawilk replied:
Could be.  I'm not seeing any other size-related fields to tweak.  I'd checked
https://nghttp2.org/documentation/nghttp2_option_set_max_send_header_block_length.html
"This option sets the maximum length of header block (a set of header fields per one HEADERS frame) to send. "
It was the "per one HEADERS frame" that got me thinking it might be the wrong field.
in code it's used in prep_frame which I thought was after the chunking but I could easily be misremembering.  I'm happy to let Auni do the code sleuthing though :-)

auni53 replied:
Yeeee, I'm pretty certain this is the right call. Also because I did this TDD-style and the testing behaviour checks out.

alyssawilk commented on test/common/http/http1/codec_impl_test.cc:
```diff
@@ -31,6 +33,7 @@ namespace Http1 {
 class Http1ServerConnectionImplTest : public TestBase {
 public:
   void initialize() {
+    http_parser_set_max_header_size(max_request_headers_kb_ * 1024);
```
Would it make sense to defer this test change to a PR when we can set this via some codec interface?  It seems strange to test the limits of http_parser if we can't actually set it in our codec.

mattklein123 replied:
+1

auni53 replied:
Instead, I was able to bump the configurable limit up so this change makes sense again.

alyssawilk commented on test/common/http/http2/codec_impl_test.cc:
```diff
@@ -913,6 +913,54 @@ TEST_P(Http2CodecImplTest, TestLargeHeadersAtLimitAccepted) {
+  request_encoder_->encodeHeaders(request_headers, true);
+}
+
+TEST_P(Http2CodecImplTest, TestSingleLargeHeadersWayOverDefaultCodecLimit) {
```
If your goal is to test an unexpectedly large header frame, you could have several keys with fairly long strings.
It makes sense to both limit the size of overall headers as well as the size of any individual header, and I think rejecting a single header key or value of 65536 bytes is totally reasonable :-)

mattklein123 commented on source/common/http/http2/codec_impl.h:
```diff
@@ -33,6 +33,7 @@ const std::string ALPN_STRING = "h2";
 // This is not the full client magic, but it's the smallest size that should be able to
 // differentiate between HTTP/1 and HTTP/2.
 const std::string CLIENT_MAGIC_PREFIX = "PRI * HTTP/2";
+static const uint32_t NGHTTP2_MAX_SEND_HEADER_BLOCK_LENGTH = 100;
```
+1, is this just for testing? I'm a little confused why we are setting this. Please add that to the comment as well. And if for testing, should this be somehow configured via the codec and only configured on our test clients?

auni53 commented:
Hi @mattklein123, working on some changes to this. I'm having an issue with the http codec configuration I could use some input on.

Basically, to set the http parser limits you need to call http_parser_set_max_header_size(x). What I'm finding from my testing is that this setting persists across test runs, i.e. making that call in one http1/codec_impl test affects the success of a different test. I had thought previously that each envoy codec would be instantiating its http parser separately, but it looks like these settings are global. Does that make sense?

mattklein123 commented:
> I had thought previously that each envoy codec would be instantiating its http parser separately, but it looks like these settings are global. Does that make sense?

Did you look at the http_parser code? I'm guessing it's setting some global variable (unfortunately).

auni53 commented:
Indeed it is. I think for now I'll make this http_set_parser call just one time, in the conn manager config. This could be messy if we ever set up per-service header limiting or something like that.

auni53 commented:
Alright, I did a bunch of cleanup here. Notable changes:

- http_parser_set_max_header_size is called by conn manager/config.cc. In my codec tests I manually set the limits on instantiation, and the limits are tested in integration test. Unfortunately I was unable to add a unit test to conn manager/config_test.cc, because this sets a static global variable my test has no way of accessing.
- I changed the expected error code to 400. This should be 431, but I need to work out how to do that. The 431 error code came from the conn manager impl rejecting the request, but now the HPE_HEADER_OVERFLOW dispatch error comes from the http_parser codec library before the conn manager checks it. I need to catch that dispatch error and reject the request at the http1 codec, but I'm not sure of what the neatest way to do that is.

mattklein123 commented on source/extensions/filters/network/http_connection_manager/config.cc:
```diff
@@ -157,6 +159,8 @@ HttpConnectionManagerConfig::HttpConnectionManagerConfig(
   route_config_provider_ = Router::RouteConfigProviderUtil::create(config, context_, stats_prefix_,
                                                                    route_config_provider_manager_);
 
+  http_parser_set_max_header_size(max_request_headers_kb_ * 1024);
```
I don't want to leak this into the HCM code. So we determined this is setting a global variable? That is really unfortunate and I don't see how we can sanely support this with different configurations. IMO we need to set some global limit for this. Compile time define for Envoy? CLI flag? Global setting? This would be the outer limit of what we can support at the per-listener config layer. Thoughts on this?

mattklein123 replied:
I thought about this more, and can we potentially just not use this library setting? We can just implement something like the HTTP/2 codec saveHeaders check in the HTTP/1 codec. Since the HTTP/1 parser is streaming, it should be pretty trivial to do our own check?

mattklein123 commented on test/common/http/http1/codec_impl_test.cc:
```diff
@@ -1,3 +1,5 @@
+#include <http_parser.h>
```
See other comment on how we expose this and set it. I think we need to rethink this.

mattklein123 replied:
Can this include be removed now?

auni53 replied:
good catch!

auni53 commented:
@mattklein123: OK, how about this. I have `DEFAULT_MAX_REQUEST_HEADERS_KB` defined in codec.h. I'll add a `MAX_CONFIGURABLE_MAX_REQUEST_HEADERS_KB`, which will be equal to the maximum setting in the conn manager proto. (If there's to infer the max proto invalidation programmatically that would be cool.)

In response to your comment on nghttp2: yes, a different codec could theoretically exceed this limit, but that can't actually happen because the hard-coded 100kb I set is higher than the max configurable (96kb) in the conn manager.

On the http1 side, I'm fine with doing the same thing and just never touching the http_parser max limit once it's been set super high. An HTTP/1-specific check is totally doable, but also maybe not even needed since the conn manager has its own check. The reason this http_parser check is coming up now is because I'm bumping the limit *up*, so if it's just by default set higher than the max configurable of 96 that's chill.

auni53 commented:
1. Cleaned up comments with KiB.

2. Set both libraries' max request headers checks to 0x7fffffff, effectively disabling them and relying on our own checks. This fixes the 431 error code problem too.

3. Remove http1/codec_impl_test.cc tests for headers being rejected, since the http1 codec now never rejects large request headers. In the integration test, http1 large headers will hit the codec_impl.cc check.

mattklein123 commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -134,13 +134,15 @@ message HttpConnectionManager {
-  // is 60K, based on default settings for http codecs. For HTTP1, the current
-  // limit set by http_parser is 80K. for HTTP2, the default allowed header
-  // block in nghttp2 is 64K. The max configurable setting is 64K in order to
+  // is 60 KiB, based on default settings for http codecs. For HTTP1, the currentB
```
typo "currentB"

mattklein123 commented on bazel/external/http-parser.BUILD:
```diff
@@ -7,6 +7,7 @@ cc_library(
         "http_parser.h",
     ],
     hdrs = ["http_parser.h"],
+    copts = ["-DHTTP_MAX_HEADER_SIZE=0x7fffffff"],
```
Can you add a comment on why we set this and why it's safe (we have checks elsewhere, etc.)

mattklein123 replied:
Also, don't you need HTTP/1 codec changes now to check the current header map size on each new header as we do with HTTP/2? I'm not sure how this is safe otherwise?

auni53 replied:
done on the comment, working on the http1 codec check

alyssawilk replied:
Also as an aside, when this lands please ping the googlers on current and next import duty as this will almost certainly need custom tweaks on import.

alyssawilk replied:
Actually sorry, on second thought I'm having a moment of paranoia.  We don't have anywhere else in the Envoy code base where we use http_parser so this is 100% OK today but what do we think of adding a limit just in case?  If we added a 200KB limit or some such it'd be very safely above our window such that we should never hit the http_parser limits in the codec, without allowing total infinite memory if someone else uses http parser.
Thoughts?

mattklein123 commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -134,13 +134,15 @@ message HttpConnectionManager {
-  // is 60K, based on default settings for http codecs. For HTTP1, the current
-  // limit set by http_parser is 80K. for HTTP2, the default allowed header
-  // block in nghttp2 is 64K. The max configurable setting is 64K in order to
+  // is 60 KiB, based on default settings for http codecs. For HTTP1, the currentB
```
Given the new changes, are these sizes still correct/relevant? Isn't is basically fully configurable now on both codecs or at least up to 96KiB max below? Can you potentially simplify the docs?

auni53 replied:
1. Simplified the docs.
2. Removed the bit on per k/v pair limits, I want to work on adding more thorough and cross-codec regression testing around that before it's documented.

mattklein123 commented on source/common/http/http2/codec_impl.h:
```diff
@@ -34,6 +34,11 @@ const std::string ALPN_STRING = "h2";
+// This setting will be passed to the nghttp2 library's max request headers check. It is set to an
+// arbitrarily high number so as to never trigger the check, as we check request headers length in
+// codec_impl::saveHeader.
+static const uint32_t NGHTTP2_MAX_SEND_HEADER_BLOCK_LENGTH = 0x7fffffff;
```
Should we not multiply by 1024 above now given we are setting this to unlimited? In fact, can you potentially just remove this and localize to the call above with a comment?

auni53 replied:
done and done

auni53 commented:
OK bless got, did the following:

1. Added back reject-case tests to http1/codec_impl_test.
2. Added a check in http1/codec_impl onHeaderValue. Lmk if there's a better spot for that, but tests pass!!!!

auni53 commented:
Wahu my tests pass, had to make some kind of clunky test interface changes (these need a bigger refactor).

alyssawilk commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -417,6 +419,13 @@ void ConnectionImpl::onHeaderValue(const char* data, size_t length) {
   }
 
+  if ((current_header_field_.size() + length + current_header_map_->byteSize()) >
+      (max_request_headers_kb_ * 1024)) {
```
If this is checking onHeaderValue in the base class, isn't it now applying to response headers (when acting as an HTTP client) as well?

I think this either needs to be renamed to indicate that it affects both request and response headers (and the default passed in above may need to be adjusted to not have a functional change) or it should only apply to the request path.

Given we plan on doing this for response headers I'd suggest doing the former rather than the latter

alyssawilk replied:
Yes, my concern is when we create 

std::make_unique<Http1::ClientConnectionImpl>(*connection_, *this, Http::DEFAULT_MAX_REQUEST_HEADERS_KB);

the client impl is sending request headers (no size checks on serialization) and receiving response headers (onHeaderValue expected to be called)

try cerr-ing in the http/1.1. integration test.  I'd expect the log to consist of the request headers received from the downstream client, and the response headers received from the upstream server.

auni53 replied:
Gotcha, fixed this and added a test.

mattklein123 commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -133,14 +133,13 @@ message HttpConnectionManager {
+  // The maximum request headers size for incoming connections.
+  // If unconfigured, the default max request headers allowed is 60 KiB.
+  // Requests that exceed this limit will receive a 431 response.
+  // The 96 KiB limit is based on bugs in nghttp2 parsing, and can be raised once those are
```
For my knowledge, what bugs are we talking about here? Either way, I would probably leave this out of the public docs and just put a comment somewhere in the code on what is going on.

auni53 replied:
Edited the comment as you asked, but I don't know what the bug is exactly. Just, some exception was thrown around 97 or 100kb of headers sent and it wasn't relevant to me to push it further.

mattklein123 commented on source/common/http/codec_client.cc:
```diff
@@ -140,7 +140,8 @@ CodecClientProd::CodecClientProd(Type type, Network::ClientConnectionPtr&& conne
   case Type::HTTP1: {
-    codec_ = std::make_unique<Http1::ClientConnectionImpl>(*connection_, *this);
+    codec_ = std::make_unique<Http1::ClientConnectionImpl>(*connection_, *this,
+                                                           Http::DEFAULT_MAX_REQUEST_HEADERS_KB);
```
I don' have a strong opinion, but I think this is probably fine. We would check response header sizes though, right? So I think this is still valid for the client?

mattklein123 commented on source/common/http/conn_manager_utility.cc:
```diff
@@ -44,8 +44,8 @@ ServerConnectionPtr ConnectionManagerUtility::autoCreateCodec(
   } else {
-    return ServerConnectionPtr{
-        new Http1::ServerConnectionImpl(connection, callbacks, http1_settings)};
+    return ServerConnectionPtr{new Http1::ServerConnectionImpl(
```
nit: as long as you are here, please change to std::make_unique

auni53 replied:
done

mattklein123 commented on source/extensions/filters/network/http_connection_manager/config.cc:
```diff
@@ -322,8 +322,8 @@ HttpConnectionManagerConfig::createCodec(Network::Connection& connection,
   case CodecType::HTTP1:
-    return Http::ServerConnectionPtr{
-        new Http::Http1::ServerConnectionImpl(connection, callbacks, http1_settings_)};
+    return Http::ServerConnectionPtr{new Http::Http1::ServerConnectionImpl(
```
nit: switch to std::make_unique

auni53 replied:
done

auni53 commented:
Okaaay so did the following:

1. max_request_headers_kb_ -> max_headers_kb_ in the ConnectionImpl.

2. ServerConnectionImpl passes stuff in the same.

3. ClientConnectionImpl gets a max_response_headers_kb_ parameter, but it's just default set to a big number. DIdn't want to expose the max_response interface in this PR without more testing around it. As well, the exception itself has to be request/response agnostic, "headers are too large", though that could be reworked later.

mattklein123 commented on source/common/http/http1/codec_impl.h:
```diff
@@ -331,8 +333,8 @@ class ServerConnectionImpl : public ServerConnection, public ConnectionImpl {
-  ClientConnectionImpl(Network::Connection& connection, ConnectionCallbacks& callbacks);
-
+  ClientConnectionImpl(Network::Connection& connection, ConnectionCallbacks& callbacks,
+                       uint32_t max_response_headers_kb = 0x7fffffff);
```
Is this really the default you want? Don't you want this limited? If this is the default, please add a comment on why.

alyssawilk replied:
If we want this PR to be a no-op for response headers, I think the response header limits should be whatever the old compiler defaults were.

alyssawilk commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -133,14 +133,11 @@ message HttpConnectionManager {
+  // Requests that exceed this limit will receive a 431 response.
   google.protobuf.UInt32Value max_request_headers_kb = 29
-      [(validate.rules).uint32.gt = 0, (validate.rules).uint32.lte = 64];
+      [(validate.rules).uint32.gt = 0, (validate.rules).uint32.lte = 96];
```
Sorry, worth commenting the reasons behind the max?

alyssawilk replied:
I think documenting somewhere that we can't go above 96 and have nghttp2 work is worthwhile.

mattklein123 replied:
Sorry to come back to this, but can you say something like "The max configurable limit is 96 KiB which is gated by the current implementation details." And then put a comment about whatever nghttp2 issues you had somewhere in the code? I don't think nghttp2 should show up in the docs, it's not relevant to the end user. If you want to put the comment here you can also put in a doc comment (see other examples) which won't get rendered.

alyssawilk replied:
+1  - that works for me as a compromise

emakarov replied:
@alyssawilk could you please describe a bit why such hard limit has been chosen?
Is there any internal implementation issues with higher limit?

mattklein123 replied:
See the comment thread above. IIRC there was some problem getting higher limits to work either in the tests or in prod code which would need debugging to allow raising the limits higher. I agree from a product perspective we should allow higher limits.

alyssawilk commented on test/common/http/http1/codec_impl_test.cc:
```diff
@@ -1022,14 +1027,88 @@ TEST_F(Http1ServerConnectionImplTest, TestCodecHeaderLimits) {
+}
+
+TEST_F(Http1ServerConnectionImplTest, TestLargeRequestHeadersAcceptedMaxConfigurable) {
+  max_request_headers_kb_ = 96;
```
shouldn't we have the same tests for response headers?

auni53 replied:
I added one test that would ascertain the response header limit behaviour doesn't change before/after this PR, but I'll change that to send 79 KiB of response headers so it's up against the vanilla 80 KiB limit.

I'll be adding more thorough response headers regression testing in my response headers PR that's wip.

alyssawilk replied:
This PR allows for codec creation with non-standard sizes.  Even if it's not configured and used end-to-end I think it's worth unit testing that the bounds you allow setting now work.

auni53 commented:
Oky I changed the ClientConnectionImpl max response headers from a default parameter to a class constant, added a test for large response headers rejection, and tweaked the API comment.

auni53 commented:
Ok, updated with Matt's suggestion on the API commenting + added a note on nghttp2 to the test file. I got a mysql_integration_test failure, hoping that was a flake bc idk how I'd have touched that.

alyssawilk commented:
Yep, that's a known and fixed flake.  If you merge master you'll pick up the fix

alyssawilk commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -419,6 +421,14 @@ void ConnectionImpl::onHeaderValue(const char* data, size_t length) {
   header_parsing_state_ = HeaderParsingState::Value;
   current_header_value_.append(data, length);
+
+  uint32_t total =
```
const uint32_t?

alyssawilk commented on source/common/http/http1/codec_impl.h:
```diff
@@ -363,6 +365,9 @@ class ClientConnectionImpl : public ClientConnection, public ConnectionImpl {
   // Set true between receiving 100-Continue headers and receiving the spurious onMessageComplete.
   bool ignore_message_complete_for_100_continue_{};
+
+  // The default limit of 80 KiB is the vanilla http_parser behaviour.
```
The default limit is 80Kb (Kilobytes) not KiB (Kibibytes), no?

alyssawilk replied:
Oh jeez, trusted the http_parser issue which was incorrect  - looks like they were saying KB when they meant KiB.  You can ignore this one.

alyssawilk commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -868,6 +868,10 @@ ConnectionImpl::Http2Options::Http2Options(const Http2Settings& http2_settings)
 
+  // The max send header block length is configured to an arbitrarily high number so as to never
+  // trigger the check within nghttp2, as we check request headers length in codec_impl::saveHeader.
+  nghttp2_option_set_max_send_header_block_length(options_, 0x7fffffff);
```
Ditto here on limits

auni53 commented:
!rekick

auni53 commented:
/retest

mattklein123 commented:
cc @RyanTheOptimist @yanavlasov @KBaichoo also on ^

RyanTheOptimist commented:
@ramaraochavali Can you file an issue about there were we can discuss the details? I think that's probably better than a 4 year old already committed PR :) Please feel free to link to this PR or any related issues there.

ramaraochavali commented:
@RyanTheOptimist  created https://github.com/envoyproxy/envoy/issues/26690
